### PR TITLE
Add SMB shares with scoped S3 access

### PR DESCRIPTION
**Pull request by Andrew-Kulpa**
## Status
**READY**

## Description
Changed the following for scoped S3 access over SMB:
- Add `var.active_directory`
- Add `var.usernames`
- Removed `var.max_size` in favor of `length(var.usernames)`
- Add `data.aws_iam_policy_document.hospital_dev_instance_access_s3` looped for each user
- Add `aws_iam_role.hospital_dev_instance_access_s3` looped for each user
- Add `aws_storagegateway_gateway.hospital_dev_s3`
- Add `aws_storagegateway_smb_file_share.hospital_dev` looped for each user

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -65,6 +65,144 @@ resource "aws_s3_bucket" "hospital_dev_s3" {
...
+
+    resources = [
+      aws_s3_bucket.hospital_dev_s3.arn
+    ]
+
+    condition {
+      test     = "StringEquals"
+      variable = "s3:prefix"
+      values = [
+        "",
```
Comment (artur-sak13) [inline main.tf]: I don't know if I've ever seen using an empty string in the prefix conditional. What exactly does it match on? "Leaf" objects?

Comment (Andrew-Kulpa) [reply main.tf]: I believe the `""` is used to implicitly specify the arn for the given bucket along with the other prefix `"home/"`. I mostly mirrored the example here: [Writing IAM Policies: Grant Access to User-Specific Folders in an Amazon S3 Bucket](https://aws.amazon.com/blogs/security/writing-iam-policies-grant-access-to-user-specific-folders-in-an-amazon-s3-bucket/)

File: variables.tf [hunk 2]
```diff
@@ -19,9 +19,31 @@ variable "vpc_cidr" {
...
 }
 
-variable "max_size" {
-  type    = number
-  default = 6
+# username parts = AWSReservedSSO_${access_level}_${id}/${SSO_identifier (email)}
+# example username =  AWSReservedSSO_AdministratorAccess_0c429936128d1c43/andrew.kulpa@twopoint.com
+variable "usernames" {
+  type = list
+  default = [
```
Comment (artur-sak13) [inline variables.tf]: Leaving a reminder to remove the `defaults` here

File: main.tf [hunk 3]
```diff
@@ -65,6 +65,144 @@ resource "aws_s3_bucket" "hospital_dev_s3" {
...
+resource "aws_iam_role" "hospital_dev_instance_access_s3" {
+  count              = length(var.usernames)
+  name               = "hospital_dev_instance_access_s3_${element(var.usernames, count.index)}"
+  assume_role_policy = element(data.aws_iam_policy_document.hospital_dev_instance_access_s3.*.json, count.index)
+}
+
+## Storage GW GW for instances
+resource "aws_storagegateway_gateway" "hospital_dev_s3" {
+  gateway_ip_address = "10.0.1.2" # TODO: dynamically set from current subnet, changing last octet
+  gateway_name       = "hospital_dev_s3_gw"
```
[the remark this answers is not in the document]
Comment (Andrew-Kulpa) [reply main.tf]: Im cool with differentiating from the resource names and using pascal_case

[the remark this answers is not in the document]
Comment (Andrew-Kulpa) [reply main.tf]: Created issue #2 for this!

File: main.tf [hunk 4]
```diff
@@ -65,6 +65,144 @@ resource "aws_s3_bucket" "hospital_dev_s3" {
...
+  statement {
+    sid    = "AllowInstanceAllS3ActionsInUserFolder"
+    effect = "Allow"
+
+    actions = [
+      "s3:*"
+    ]
+
+    resources = [
+      "${aws_s3_bucket.hospital_dev_s3.arn}/home/AWSReservedSSO_*_*/${element(var.usernames, count.index)}@twopoint.com/*",
```
[the remark this answers is not in the document]
Comment (Andrew-Kulpa) [reply main.tf]: I think we set them using ARNs according to the [example](https://aws.amazon.com/blogs/security/writing-iam-policies-grant-access-to-user-specific-folders-in-an-amazon-s3-bucket/) I followed but I am all good with refactoring to bucket names if thats possible!

[the remark this answers is not in the document]
Comment (artur-sak13) [reply main.tf]: Yep! You're right, my bad.

File: variables.tf (hunk 2 above)
Comment (Andrew-Kulpa) [reply variables.tf]: /reminder Artur I put the defaults reminder as a `TODO:` comment

File: main.tf [hunk 5]
```diff
@@ -65,6 +65,144 @@ resource "aws_s3_bucket" "hospital_dev_s3" {
...
+  )
+}
+
+## SMB File Shares for each Instance
+resource "aws_storagegateway_smb_file_share" "hospital_dev" {
+  count          = length(var.usernames)
+  authentication = "ActiveDirectory"
+  gateway_arn    = aws_storagegateway_gateway.hospital_dev_s3.arn
+  location_arn   = aws_s3_bucket.hospital_dev_s3.arn
+  role_arn       = element(aws_iam_role.hospital_dev_instance_access_s3.*.arn, count.index)
```
[the remark this answers is not in the document]
Comment (Andrew-Kulpa) [reply main.tf]: Yep, good catch! Pushed this recently 👍

**Pull request merged**